libc/dirent/opendir.c

```diff
@@ -0,0 +1,447 @@
+/*
+ * Phoenix-RTOS
+ *
+ * libphoenix
+ *
+ * test/test_strftime.c
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Arkadiusz Kozlowski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <time.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <dirent.h>
+#include <limits.h>
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <string.h>
+
+#include <unity_fixture.h>
+
+#include "common.h"
+
+#define MAIN_DIR "test_opendir"
+
+
+int create_directories(int num_of_dirs, char identifier)
+{
+
+	char dirPath[40];
+	DIR *dirs[num_of_dirs];
+	int opened_dirs = 0;
+	int result = 0;
+
+
+	/* Create directories in batch */
+	for (int i = 0; i < num_of_dirs; ++i) {
+
+		sprintf(dirPath, MAIN_DIR "/%d%c", i, identifier);
+
+		if (mkdir(dirPath, 0777) == -1 && errno != EEXIST) {
+			TEST_FAIL_MESSAGE("Directory creating error");
+		}
+	}
+
+	/* Open directories one by one, until one of them fails */
+	for (int i = 0; i < num_of_dirs; ++i) {
+
+		sprintf(dirPath, MAIN_DIR "/%d%c", i, identifier);
+
+		/*
+		 * Guard clause that skips current take if dir was opened,
+		 * Upon failing it proceeds to cleanup
+		 */
+		if (dirs[i] = opendir(dirPath)) {
+			opened_dirs = i;
+			continue;
+		}
+
+		result = -1;
+		break;
+	}
+
+	for (int i = 0; i <= opened_dirs; ++i) {
+		TEST_ASSERT_EQUAL(0, closedir(dirs[i]));
+	}
+
+	for (int i = 0; i < num_of_dirs; ++i) {
+		sprintf(dirPath, MAIN_DIR "/%d%c", i, identifier);
+		rmdir(dirPath);
+	}
+
+	return result;
+}
+
+
+TEST_GROUP(dirent_opendir);
+
+
+TEST_SETUP(dirent_opendir)
+{
+
+	mkdir(MAIN_DIR, 0777);
+	mkdir(MAIN_DIR "/dir_without_read_perm", 0000);
+	FILE *fptr;
+	if (fptr = fopen(MAIN_DIR "/notadir.txt", "w")) {
+		fprintf(fptr, "Some file contents");
+		fclose(fptr);
+	}
+}
+
+
+TEST_TEAR_DOWN(dirent_opendir)
+{
+	errno = 0;
+	remove(MAIN_DIR "/notadir.txt");
+	chmod(MAIN_DIR "/dir_without_read_perm", 0777);
+	rmdir(MAIN_DIR "/dir_without_read_perm");
+	rmdir(MAIN_DIR);
+}
+
+
+TEST(dirent_opendir, opening_empty_directory)
+{
+	DIR *dp;
+	mkdir(MAIN_DIR "/empty_dir", 0777);
+	TEST_ASSERT_NOT_NULL(dp = opendir(MAIN_DIR "/empty_dir"));
+	closedir(dp);
+	rmdir(MAIN_DIR "/empty_dir");
+}
+
+
+TEST(dirent_opendir, opening_not_empty_directory)
+{
+	DIR *dp;
+	TEST_ASSERT_NOT_NULL(dp = opendir(MAIN_DIR));
+	closedir(dp);
+}
+
+
+TEST(dirent_opendir, no_read_permission)
+{
+	char unreadable[] = MAIN_DIR "/dir_without_read_perm";
+	char readable[] = MAIN_DIR "/dir_without_read_perm/readable_dir";
+	DIR *dirPtr;
+
+
+	chmod(unreadable, 0700);
+	mkdir(readable, 0777);
+	chmod(unreadable, 0000);
```

Comment:
Changing permission to directories which not exist, if you checked return value of `chmod()` you'd have got it.

Reply:
It exists, it's just created in TEST_SETUP.